### DISCUSSION TITLE
Linkerd Sidecar never exits on completed job pod?

### DISCUSSION THREAD

**Discussion by jonstelly** [linkerd/linkerd2#6436] [Q&A]
I've deployed linkerd using the CLI and turned on injection for my default namespace.  It seems like the linkerd sidecar stays active even after the primary container for the pod reaches the completed state, which means that the job never registers that the pod has completed.

If I disable linkerd for that job/pod, then it works as expected, but it would be nicer if I didn't have to tweak every helm chart by disabling linkerd injection.  Is there some linkerd setting to make the sidecar exit when all other containers in a pod have completed?

**Comment by mateiidavid** [comment]
Hey @jonstelly, 

This is a good question. In order to terminate, the linkerd-proxy needs to receive a SIGTERM signal. Normally, when a pod is terminated (for whatever reason), k8s will send a termination signal to all containers in the pod. When a Job object wraps a pod, however, the pod terminates only when all containers terminate successfully -- in other words, k8s won't send it a SIGTERM signal, so you'd have to do it yourself. If you want to read more about it, you can read about our docs on "[Graceful Pod Shutdown](https://linkerd.io/2.10/tasks/graceful-shutdown/)".

There are two options that I can personally think of here, both would involve slightly changing your code/pipeline to accommodate this. There might be other ways to do it, but I'm just not familiar with them.
  * **(1)**: In [linkerd/linkerd-proxy/PR#811](https://github.com/linkerd/linkerd2-proxy/pull/811) we introduced a `/shutdown` endpoint for all proxies on their admin port (`4191` by default). You can send a manually signal termination from your main container; since all containers run in the same networking namespace, you can (and *should*) make the request over the loopback interface. For example, if you're using Go, you can add a `defer sendShutdownSignal()` in your main function, the shutdown function in this case will send a request to `http://127.0.0.1:4191/shutdown` and that should terminate the proxy container.
  * **(2)**: Use [linkerd/linkerd-await](https://github.com/linkerd/linkerd-await). `linkerd-await` wraps your application and is generally used to block the main container until the proxy is ready to accept traffic. You can, however, also use it to send a shutdown signal when the main container exits. There's a `--shutdown` flag that you can use, the README should be pretty descriptive. The downside is, you have to change your Docker image to include `linkerd-await`; this will be your new entrypoint. To send a shutdown signal, it actually uses method (1), so if you want some inspiration you can have a look at [this function in the linkerd-await codebase](https://github.com/linkerd/linkerd-await/blob/main/src/main.rs#L183). 

Hope this all makes sense, let me know if you have any questions! :)

**Reply by jonstelly** [reply]
Thanks for that detailed response.  I've been digging through all of the challenges around sidecars on K8s.  Given that [the KES to improve sidecars in jobs won't be progressing](https://github.com/kubernetes/enhancements/issues/753#issuecomment-713471597), It seems like [this](https://github.com/linkerd/linkerd2/issues/1869#issuecomment-595456178) may be one more option.

But it seems like those options are pretty challenging to apply to jobs that come from 3rd party projects, some helm charts have jobs that don't expose any customization / values support for this, sometimes those charts will support annotations or labels on the jobs and that's it.

Would it be reasonable to add another linkerd annotation, something like `linkerd.io/auto-shutdown: enabled`, where the proxy could watch for other container processes using either `shareProcessNamespace` like mentioned in my link, or maybe through [the downward api](https://kubernetes.io/docs/tasks/inject-data-application/downward-api-volume-expose-pod-information/)?

The easiest option here (for linkerd users) would be to support that at a namespace and have the linkerd proxy injector add necessary mutations for job pods.

Or if that doesn't work, maybe an attribute like `linkerd.io/disable-on-jobs: true` that prevents linkerd injection for job pods?

Sorry if these are silly questions, I only have a very basic understanding of mutating webhooks.

EDIT: Alternative ideas:

`linkerd.io/inject: enabled:-jobs` - So inject could be set at a namespace level but jobs would be ignored by the proxy injector
`linkerd.io/watch-container: my-container` - To get container status out of podinfo and identify the primary container, where linkerd proxy exits X seconds after the primary container completes

I'm not a go developer but I'd be happy to help out with implementing and testing if any of these or other methods would be workable.

**Reply by mateiidavid** [reply]
@jonstelly thanks for the research on this! Bit unfortunate that the sidecar KES will not be progressing, we were all looking forward to better support there.

The options you listed are good; there's an alternative way to do this by using k8s container hooks (as opposed to using the `shareProcessNamespace`). We managed to add [support to await proxy readiness by default](https://github.com/linkerd/linkerd2/pull/5967) using this method. The pull request that I linked has all of the details, but in essence, we managed to do this by packaging `linkerd-await` as a binary in the proxy's dockerfile and having the proxy-injector re-order the containers in the pod spec. The proxy container comes up first, and since it has a `postHook` [[1]], `linkerd-await` is called before the next container comes up (essentially blocking the main container from starting until await completes). I wrote all of this down because you've been very thorough with your investigation and I thought this might be interesting to go through.

To go back on topic, we'd prefer to keep Linkerd's configuration minimal -- our philosophy is that Linkerd should easy to use and it should limit configuration options -- as such, we'd prefer not to add a special case to the proxy injector for `Job` types and we'd prefer not to introduce a new annotation. Default support for awaiting proxy readiness was a bit of an exception to this rule since many types of applications require to block the main container until the proxy is up (and ready to handle traffic). This particular issue though is very specific to `Job` types hence the hesitation in adding more configuration to the project. With that being said, I'm happy to turn this discussion into an actual GitHub issue to get more opinions and to continue the conversation, if you'd like.

In my opinion, the preferred approach here would be to have an extension centred around adding `linkerd-await` (with various configuration options) to injected workloads. This could be a simple `MutatingWebhookConfiguration` server that would function in a similar way to the proxy-injector; based on annotations it could add `linkerd-await` and re-order the containers in the pod spec. As an extension, configuration would be decoupled from the core control plane components and it would be something that people can opt in to. If you'd like to have a weekend project around this and build an extension, there are a couple of examples I can point you to (and of course we can help you out with any questions you might have along the way).

> Or if that doesn't work, maybe an attribute like linkerd.io/disable-on-jobs: true that prevents linkerd injection for job pods?

You can already do this by adding `linkerd.io/inject: disabled` to your Job's pod template. This should overwrite any namespace configuration you have in place, you can disable Linkerd on a workload-by-workload basis [[2]]. Might be an avenue worth exploring if you don't need your Job to be part of the mesh.

Let me know what you think. If you'd like this turned into an issue, we can also move the conversation there.

[1]: https://kubernetes.io/docs/tasks/configure-pod-container/attach-handler-lifecycle-event/
[2]: https://linkerd.io/2.10/features/proxy-injection/#overriding-injection

**Reply by jonstelly** [reply]
> You can already do this by adding linkerd.io/inject: disabled to your Job's pod template

True, one of the things I was hoping for was an attribute that could be added at the namespace level like `inject` so I wouldn't have to tweak every helm chart I deploy that uses a job, sometimes those charts don't even have annotation/label support for the jobs (I remember running into this somewhere but can't remember the specific chart(s)).

I like the idea of another mutating webhook, as I understand it those things just get called one after another so keeping this behavior code completely separate from the core linkerd behavior makes complete sense.

I'll take a look at the PR you linked about awaiting proxy readiness and if I get some time I may try my hand at the webhook extension.

No need to create an issue for this, since you were so thorough I imagine people searching for linkerd + jobs will find their way here.

Thanks again and thanks for linkerd!  It's one of my favorite kubernetes things, especially the linkerd multicluster support which I wasn't even looking for but just happened to be a great solution for federating prometheus across clusters.
